### ISSUE TITLE
ObjectDisposedException from HttpContent.LoadIntoBufferAsync with WindowsAuth

### ISSUE THREAD

**Issue by Tratcher** [dotnet/corefx#32143]
Starting Sept 2nd ASP.NET Core started seeing unusual ODEs from HttpClient. There have been 26 automated repros in the last four days. 2.2 and 3.0 both reproduce the issue. I have not been able to reproduce it manually in 2.2, but it fails every time in 3.0. https://github.com/aspnet/HttpSysServer/issues/474

In the 3.0 branch the failures started after updating to .NET Core 3.0.0-preview1-26905-03.

```
System.ObjectDisposedException : Cannot access a disposed object.
Object name: 'System.Net.Http.HttpConnection+HttpConnectionResponseContent'.
   at System.Net.Http.HttpContent.CheckDisposed()
   at System.Net.Http.HttpContent.LoadIntoBufferAsync(Int64 maxBufferSize, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
   at Microsoft.AspNetCore.Server.HttpSys.AuthenticationTests.SendRequestAsync(String uri, Boolean useDefaultCredentials) in /_/test/Microsoft.AspNetCore.Server.HttpSys.FunctionalTests/AuthenticationTests.cs:line 377
   at Microsoft.AspNetCore.Server.HttpSys.AuthenticationTests.AuthTypes_UnathorizedAuthenticatedAuthType_Unauthorized(AuthenticationSchemes authType) in /_/test/Microsoft.AspNetCore.Server.HttpSys.FunctionalTests/AuthenticationTests.cs:line 364
--- End of stack trace from previous location where exception was thrown ---
```

The scenario is Windows Auth Negotiate tests with HttpSys server. The NTLM variant is not impacted. There have been no server side changes here for some time.
https://github.com/aspnet/HttpSysServer/blob/9cb675fec89702d28b5f188e0ab4f7d671d7aa11/test/Microsoft.AspNetCore.Server.HttpSys.FunctionalTests/AuthenticationTests.cs#L349-L369

The client code is not doing anything interesting that should lead to an ODE:
https://github.com/aspnet/HttpSysServer/blob/9cb675fec89702d28b5f188e0ab4f7d671d7aa11/test/Microsoft.AspNetCore.Server.HttpSys.FunctionalTests/AuthenticationTests.cs#L371-L379

The test fails in only 1 second so timeouts should not be a factor. Nothing else the server does should be capable of causing an ODE here on the client.

I see some recent changes in this code path on the client that may be related: https://github.com/dotnet/corefx/pull/31589/files

@davidsh

**Comment by davidsh** [comment]
@geoffkizer

**Comment by geoffkizer** [comment]
@Tratcher How do I run the repro?

**Comment by Tratcher** [comment]
Alternate repro:

Enable Windows Auth in IIS and disable Anonymous. Oddly in IIS you need to make sure Kernel Mode auth is disabled (see Windows Auth Advanced Settings), whereas HttpSys only uses Kernel mode auth.
```
        public static void Main(string[] args)
        {
            HttpClientHandler handler = new HttpClientHandler();
            handler.UseDefaultCredentials = true;
            using (HttpClient client = new HttpClient(handler))
            {
                var result = client.GetAsync("http://localhost:80").Result;
                Console.WriteLine(result);
            }
        }
```

**Comment by geoffkizer** [comment]
I can't repro with your alternate repro. Works fine on my machine.

**Comment by geoffkizer** [comment]
I tried with both kernel auth enabled and kernel auth disabled, and it worked both ways.

That said, looking at the listed auth providers, I seem to only have NTLM enabled. I'm guessing you have Negotiate enabled as well? I have a bunch of different Negotiate providers listed here, none of which are enabled. Can you tell me which ones you have enabled?

**Comment by geoffkizer** [comment]
I enabled Negotiate (and put it first in the provider list) and it still works, with or without kernel auth enabled.

**Comment by geoffkizer** [comment]
Do you have a simple httpsys repro for this? Might be easier than fiddling with IIS config.

**Comment by Tratcher** [comment]
[WinAuthODE.zip](https://github.com/dotnet/corefx/files/2358588/WinAuthODE.zip)
I've attached a server that reproduces the issue against the above client code. This server app is written using 2.0. Run the WebApplication13 profile rather than IIS Express.

I realized this test does do one unusual thing, it sends a 401 again after the successful login to indicate the user did not permissions to access the resource. I don't know why I could repro the ODE against IIS without this addition and you couldn't.

**Comment by geoffkizer** [comment]
It looks like we broke the case where the server still sends 401 after auth completes. We are draining the response in this case when we shouldn't be.

I'll get a fix ready.

@karelz @davidsh We will want to backport this as well.

**Comment by geoffkizer** [comment]
Regression was introduced here: https://github.com/dotnet/corefx/pull/31589

**Comment by davidsh** [comment]
>The scenario is Windows Auth Negotiate tests with HttpSys server. The NTLM variant is not impacted. 

@Tratcher @geoffkizer I don't understand why this only repro's with Negotiate and not NTLM.  Have this been figured out yet?  The PR recently opened to fix this regression doesn't seem Negotiate-specific at all.

Perhaps this does repro with both Negotiate and NTLM now that it has been determined to be an problem regarding a final 401 response code.

**Comment by geoffkizer** [comment]
I agree, it seems like it should repro with both. I didn't try NTLM, only negotiate.

**Comment by Tratcher** [comment]
Nor do I. Does NTLM send different headers on the final response that result in a connection closure and a different HttpContent type?

**Comment by davidsh** [comment]
>I realized this test does do one unusual thing, it sends a 401 again after the successful login to indicate the user did not permissions to access the resource. 

@Tratcher How do I modify the server repro you posted to return 200 instead of 401?

**Comment by Tratcher** [comment]
In Startup.cs change
```
            app.Run(context =>
            {
                return context.ChallengeAsync(HttpSysDefaults.AuthenticationScheme, null);
            });
```
to
```
            app.Run(context =>
            {
                return Task.CompletedTask;
            });
```

**Comment by davidsh** [comment]
Actually, I want it to do the authentication dance with NTLM or Negotiate.  I just want it to end up with 401.  So, does the change above you suggested still do that?

**Comment by Tratcher** [comment]
The authentication dance will still be managed by the server as configured in Main. This call only controls the last response so yes you will get auth and then 200.

**Comment by davidsh** [comment]
>Nor do I. Does NTLM send different headers on the final response that result in a connection closure and a different HttpContent type?

No.  I still can't figure out why it repros with Negotiate and not NTLM. But I am studying the WireShark traces.

In both cases, the final 401 response is the same (except for the NTLM vs Negotiate token):

Negotiate:

```
HTTP/1.1 401 Unauthorized
Content-Length: 0
Server: Microsoft-HTTPAPI/2.0
WWW-Authenticate: Negotiate
Date: Fri, 07 Sep 2018 22:47:00 GMT
```

NTLM:

```
HTTP/1.1 401 Unauthorized
Content-Length: 0
Server: Microsoft-HTTPAPI/2.0
WWW-Authenticate: NTLM
Date: Fri, 07 Sep 2018 22:44:59 GMT
```

**Comment by davidsh** [comment]
And the PR #32146 fixes both `break` code paths because it ensures that the response is only drained if there is going to be a subsequent request to send an additional Authorization header blob.

**Comment by davidsh** [comment]
>I enabled Negotiate (and put it first in the provider list) and it still works, with or without kernel auth enabled.

I also have a theory why this is difficult to repro in IIS.  As I mentioned above, this only repro's if a particular `break` causes the `while(true)` loop to exit.  Even though you might be using Negotiate scheme, it still might end up doing NTLM underneath.  And that will cause the same behavior as NTLM scheme itself.  In the cases where you think it should repro even though you specified Negotiate scheme, if you look at the WireShark traffic, you can see that it is really NTLM and not Kerberos protocol.

NTLM traffic has base-64 blob starting with "T".  And Kerberos traffic has base-64 blobs starting with "Y".  Kerberos is only used if all machines are domain-joined and other requirements are met.  So, that might explain some weirdness in the repro scenario since the repro must hit the particular `break` line in the while loop.

**Comment by davidsh** [comment]
Fixed in master with #32146.

**Comment by davidsh** [comment]
Fixed in release/2.1 branch (for 2.1.5) with #32243
